// @axiom-crypto/core@0.3.6/core/axiom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodePackedQuery = exports.Axiom = void 0;
const internalConfig_1 = require("./internalConfig");
const block_1 = require("./block");
const query_1 = require("./query");
const queryBuilder_1 = require("../query/queryBuilder");
const abi_1 = require("./lib/abi");
class Axiom {
    constructor(config, overrides) {
        this.config = new internalConfig_1.InternalConfig(config, overrides);
        this.block = new block_1.Block(this.config);
        this.query = new query_1.Query(this.config);
    }
    newQueryBuilder() {
        return new queryBuilder_1.QueryBuilder(this.config);
    }
    getAxiomAbi() {
        return (0, abi_1.getAxiomAbiForVersion)(this.config.version);
    }
    getAxiomQueryAbi() {
        return (0, abi_1.getAxiomQueryAbiForVersion)(this.config.version);
    }
    getAxiomAddress() {
        return this.config.getConstants().Addresses.Axiom;
    }
    getAxiomQueryAddress() {
        return this.config.getConstants().Addresses.AxiomQuery;
    }
}
exports.Axiom = Axiom;
var decoder_1 = require("../query/decoder");
Object.defineProperty(exports, "decodePackedQuery", { enumerable: true, get: function () { return decoder_1.decodePackedQuery; } });
